### PR TITLE
Support block scoping

### PR DESCRIPTION
**Pull request by arsalan0c**
This PR adds support for block scoping along with supporting test code which results in an error.

### CODE REVIEW

File: src/__tests__/non-det-interpreter.ts [hunk 1]
```diff
@@ -208,32 +208,130 @@ test('Require operator', async () => {
...
+export async function testNonDeterministicCode(
+  code: string,
+  expectedValues: any[],
+  hasError?: boolean
+) {
   const context = makeNonDetContext()
   let result: Result = await runInContext(code, context, nonDetTestOptions)
-  const numOfRuns = expectedValues.length
+
+  const numOfRuns = hasError ? expectedValues.length - 1 : expectedValues.length
```
Comment (anubh-v) [inline src/__tests__/non-det-interpreter.ts:322]: Why not `  const numOfRuns = hasError ?  0: expectedValues.length` ?

File: src/__tests__/non-det-interpreter.ts [hunk 2]
```diff
@@ -208,32 +208,130 @@ test('Require operator', async () => {
...
-  expect(result.status).toEqual('finished')
-  expect((result as Finished).value).toEqual(undefined)
+  if (!hasError) {
+    // all non deterministic programs have a final result whose value is undefined
+    expect(result.status).toEqual('finished')
+    expect((result as Finished).value).toEqual(undefined)
+  } else {
+    expect(result.status).toEqual('error')
+    const message: string = parseError([(result as Error).message as SourceError])
+    expect(message).toEqual(expectedValues[expectedValues.length - 1])
```
Comment (anubh-v) [inline src/__tests__/non-det-interpreter.ts:337]: why not expectedValues[0] ?

File: src/__tests__/non-det-interpreter.ts [hunk 3]
```diff
@@ -208,32 +208,130 @@ test('Require operator', async () => {
...
+  await testDeterministicCode(
+    `
+    const a = 1;
+    function f() {
+      display(a);
+      const a = 5;
+    }
+    f();
+    `,
+    'Line -1: Name a declared later in current scope but not yet assigned',
```
Comment (anubh-v) [inline src/__tests__/non-det-interpreter.ts:265]: Why is this Line -1 ?

File: src/interpreter/interpreter-non-det.ts [hunk 4]
```diff
@@ -436,14 +436,18 @@ export const evaluators: { [nodeType: string]: Evaluator<es.Node> } = {
...
-    const environment = createBlockEnvironment(context, 'blockEnvironment')
+    let environment = createBlockEnvironment(context, 'blockEnvironment')
     pushEnvironment(context, environment)
-    result = yield* evaluateBlockSatement(context, node)
+    const resultGenerator = evaluateBlockSatement(context, node)
+
+    for (const result of resultGenerator) {
+      popEnvironment(context)
+      yield result
+      environment = createBlockEnvironment(context, 'blockEnvironment')
```
Comment (anubh-v) [inline src/interpreter/interpreter-non-det.ts]: Shouldn't we re-use the same environment created in line 440?

Comment (arsalan0c) [reply src/interpreter/interpreter-non-det.ts]: Yes that should be so.

File: src/__tests__/non-det-interpreter.ts (hunk 3 above)
Comment (arsalan0c) [reply src/__tests__/non-det-interpreter.ts:265]: See #29 - I'm not sure of the reason yet.

File: src/__tests__/non-det-interpreter.ts (hunk 1 above)
Comment (arsalan0c) [reply src/__tests__/non-det-interpreter.ts:322]: For non-deterministic tests, the error does not have to occur on the first run.

File: src/__tests__/non-det-interpreter.ts (hunk 2 above)
Comment (arsalan0c) [reply src/__tests__/non-det-interpreter.ts:337]: See above.

File: src/__tests__/non-det-interpreter.ts (hunk 3 above)
Comment (anubh-v) [reply src/__tests__/non-det-interpreter.ts:265]: Thanks for creating the issue

File: src/__tests__/non-det-interpreter.ts (hunk 1 above)
Comment (anubh-v) [reply src/__tests__/non-det-interpreter.ts:322]: I see, yes that makes sense.

File: src/types.ts [hunk 5]
```diff
@@ -146,6 +146,7 @@ export interface Environment {
 
 export interface Error {
   status: 'error'
+  message?: SourceError
```
Comment (anubh-v) [inline src/types.ts]: Hmm..would it be possible not to modify the Error type? Since it's used by other components too

Comment (anubh-v) [reply src/types.ts]: I think the Context has an array of SourceErrors right? Can we access the SourceError  from the Context?
If you look at the `testInContext` method in `testing.ts`, it does call `parseInError` using the errors' array in Context

Comment (arsalan0c) [reply src/types.ts]: Yeah that works!

**Review by anubh-v — APPROVED** [review]

**Pull request merged**